server/src/updaters/relationship-updaters.js

```diff
@@ -0,0 +1,202 @@
+
+async function updateRelationship(
+  viewer: Viewer,
+  request: RelationshipRequest,
+) {
+  const { userID, status } = request;
+
+  if (viewer.id === userID) {
+    throw new ServerError('invalid_user');
+  }
+
+  const userQuery = SQL`SELECT 1 FROM users WHERE id = ${userID} LIMIT 1`;
+  const [users] = await dbQuery(userQuery);
+
+  if (!users.length) {
+    throw new ServerError('user_does_not_exist');
+  }
+
+  // friend request
+  if (status === relationshipStatuses.PENDING_FRIEND) {
+    const [relationships] = await fetchRelationshipQuery(viewer.id, userID);
+
+    if (!relationships.length) {
+      await createRelationshipQuery(viewer.id, userID, status);
+    } else {
+      const [actorRelationship] = getRelationshipsForActorUser(
+        relationships,
+        viewer.id,
+      );
+      if (actorRelationship) {
+        const message = actorFriendshipErrorMap[actorRelationship.status];
+        throw new ServerError(message);
+      }
+
+      const [receiverRelationship] = getRelationshipsForReceiverUser(
+        relationships,
+        viewer.id,
+      );
+      if (receiverRelationship) {
+        const message = receiverFriendshipErrorMap[receiverRelationship.status];
+        throw new ServerError(message);
+      }
+    }
+    // block action
+  } else if (status === relationshipStatuses.BLOCKED) {
+    const [relationships] = await fetchRelationshipQuery(viewer.id, userID);
+
+    if (!relationships.length) {
+      await createRelationshipQuery(viewer.id, userID, status);
+    } else {
+      const [actorRelationship] = getRelationshipsForActorUser(
+        relationships,
+        viewer.id,
+      );
+      if (actorRelationship) {
+        if (isPendingFriendOrFriend(actorRelationship.status)) {
+          const { user1, user2 } = actorRelationship;
+          // cancels already sent friend request or removes friendship then blocks the user
+          await updateRelationshipQuery(user1, user2, status);
+        }
+
+        if (actorRelationship.status === relationshipStatuses.BLOCKED) {
+          throw new ServerError('user_already_blocked');
+        }
+      }
+
+      const [receiverRelationship] = getRelationshipsForReceiverUser(
+        relationships,
+        viewer.id,
+      );
+      if (receiverRelationship) {
+        if (isPendingFriendOrFriend(receiverRelationship.status)) {
+          const { user1, user2 } = receiverRelationship;
+          // cancels already received friend request or removes frienship then blocks user
+          await deleteRelationshipQuery(user1, user2);
+          await createRelationshipQuery(viewer.id, userID, status);
+        }
+
+        if (receiverRelationship.status === relationshipStatuses.BLOCKED) {
+          // users blocked each other
+          await createRelationshipQuery(viewer.id, userID, status);
+        }
+      }
+    }
+    // friend request acceptance
+  } else if (status === relationshipStatuses.FRIEND) {
+    const query = SQL`
```

Comment:
I don't think we need a separate query for this, can we just reuse `fetchRelationshipQuery`?

Reply:
I think we can. I need to query `PENDING_FRIEND` relationship initiated by a viewer, but I think I can use `fetchRelationshipQuery` then filter it out in memory